types/got/index.d.ts

```diff
@@ -98,6 +87,37 @@ declare namespace got {
         (url: GotUrl, options: GotBodyOptions<null>): GotPromise<Buffer>;
     }
 
+    interface GotJSONFn {
+        (url: GotUrl): GotPromise<any>;
+        (url: GotUrl, options: GotJSONOptions): GotPromise<any>;
+    }
+
+    type GotInstance<T = GotFn> = T &
+        Record<'get' | 'post' | 'put' | 'patch' | 'head' | 'delete', T> &
+        {
+            stream: GotStreamFn & Record<'get' | 'post' | 'put' | 'patch' | 'head' | 'delete', GotStreamFn>;
+            extend: GotExtend;
+            RequestError: typeof RequestError;
+            ReadError: typeof ReadError;
+            ParseError: typeof ParseError;
+            HTTPError: typeof HTTPError;
+            MaxRedirectsError: typeof MaxRedirectsError;
+            UnsupportedProtocolError: typeof UnsupportedProtocolError;
+            CancelError: typeof CancelError;
+            TimeoutError: typeof TimeoutError;
+        };
+
+    type DefaultGotInstance = GotInstance;
+    type JSONGotInstance = GotInstance<GotJSONFn>;
+
+    interface GotExtend {
+        (options: GotJSONOptions): JSONGotInstance;
+        (options: GotFormOptions<string>): DefaultGotInstance;
```

Comment:
For type safety need to return specific instances for each overload, not only for JSON, e.g.:

```ts
interface GotExtend {
    (options: GotFormOptions<string>): GotInstance<GotFormFn<string>>;
    (options: GotFormOptions<null>): GotInstance<GotFormFn<null>>;
}
```

Reply:
Think I've done this now, though I'm not entirely sure what these types actually do, so... ¯\\\_(ツ)_/¯

Reply:
The main idea of `GotFn` interface is allow to match parameters with resulting type, e.g.:

```ts
const result: GotPromise<Buffer> = got('http://example.com', { 
    form: true, 
    encoding: null 
}); 
```
In the example above response body has `Buffer` type (not `string` or `any`), because parameters `form` and `encoding` were specified.

For `got.extend` similar approach can be used:

```ts
const client = got.extend({ form: true, encoding: null });
const result: GotPromise<Buffer> = client('http://example.com'); 
```

Otherwise, if `got.extend()` returns `GotInstance` for `GotFormOptions<null>` parameters, resulting body will have `string` type, which is incorrect.

Reply:
Thanks for the extra info, I've added some tests for those scenarios and fixed the types :)

I've only ever used got with JSON APIs, so these types have always just seemed... Pretty strange 😂